packages/go_router/lib/src/state.dart

```diff
@@ -82,3 +82,71 @@ class GoRouterState {
         params: params, queryParams: queryParams);
   }
 }
+
+/// The current state for a [StatefulShellRoute].
+class StatefulShellRouteState {
+  /// Constructs a [StatefulShellRouteState].
+  StatefulShellRouteState({
+    required this.route,
+    required this.navigationBranchState,
+    required this.index,
+  });
+
+  /// The associated [StatefulShellRoute]
+  final StatefulShellRoute route;
+
+  /// The state for all separate route branches associated with a
+  /// [StatefulShellRoute].
+  final List<ShellRouteBranchState> navigationBranchState;
+
+  /// The index of the currently active route branch.
+  final int index;
+
+  /// Gets the current location from the [topRouteState] or falls back to
+  /// the root path of the associated [route].
+  String get location => navigationBranchState[index].location;
+
+  /// Gets the [Navigator]s for each of the route branches. Note that the
+  /// Navigator for a particular branch may be null if the branch hasn't been
+  /// visited yet.
+  List<Widget?> get navigators => navigationBranchState
+      .map((ShellRouteBranchState e) => e.navigator)
+      .toList();
+}
+
+/// The current state for a particular route branch
+/// ([ShellRouteBranch]) of a [StatefulShellRoute].
+class ShellRouteBranchState {
+  /// Constructs a [ShellRouteBranchState].
+  ShellRouteBranchState({
+    required this.navigationItem,
+    required this.rootRoutePath,
+    this.navigator,
+    this.topRouteState,
+  });
+
+  /// The associated [ShellRouteBranch]
+  final ShellRouteBranch navigationItem;
+
+  /// The full path at which root route for the route branch is reachable.
+  final String rootRoutePath;
+
+  /// The [Navigator] for this route branch in a [StatefulShellRoute]. This
+  /// field will typically not be set until this route tree has been navigated
+  /// to at least once.
+  final Navigator? navigator;
```

Comment:
No, this needs to be public to enable builder implementations using a custom container for the Navigators (i.e. not the IndexStack). In this scenario, the Navigator will not be available via the globalkey, since it hasn't been added to the widget tree yet.